File: UD-04/P02-Encordados-MongoDB-JPA-Java/src/main/java/es/joseluisgs/p02encordadosmongodbjpa/controllers/MutuaController.java
package es.joseluisgs.p02encordadosmongodbjpa.controllers;

import es.joseluisgs.p02encordadosmongodbjpa.models.Raqueta;
import es.joseluisgs.p02encordadosmongodbjpa.models.Tenista;
import es.joseluisgs.p02encordadosmongodbjpa.repositories.RaquetasRepository;
import es.joseluisgs.p02encordadosmongodbjpa.repositories.TenistasRepository;
import org.bson.types.ObjectId;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;

import java.util.List;
import java.util.Optional;
import java.util.logging.Logger;

@Controller
public class MutuaController {
    private final Logger logger = Logger.getLogger(MutuaController.class.getName());

    // Mis dependencias
    private final RaquetasRepository raquetasRepository;
    private final TenistasRepository tenistasRepository;

    @Autowired
    public MutuaController(RaquetasRepository raquetasRepository, TenistasRepository tenistasRepository) {
        this.raquetasRepository = raquetasRepository;
        this.tenistasRepository = tenistasRepository;
    }

    // Tenistas
    public void tenistasDeleteAll() {
        logger.info("Borrando todos los tenistas");
        tenistasRepository.deleteAll();
    }

    public List<Tenista> getTenistas() {
        logger.info("Obteniendo Tenistas");
        return tenistasRepository.findAll();
    }

    public Tenista createTenista(Tenista tenista) {
        logger.info("Creando Tenista");
        return tenistasRepository.save(tenista);
    }

    public Optional<Tenista> getTenistaById(ObjectId id) {
        logger.info("Obteniendo Tenista con id: " + id);
        return tenistasRepository.findById(id);
    }

    public Tenista updateTenista(Tenista tenista) {
        logger.info("Actualizando Tenista con uuid: " + tenista.getId());
        return tenistasRepository.save(tenista);
    }

    public void deleteTenista(Tenista tenista) {
        logger.info("Eliminando Tenista con id: " + tenista.getId());
        tenistasRepository.delete(tenista);
    }

    // Raquetas
    public void raquetasDeleteAll() {
        logger.info("Borrando todas los raquetas");
        raquetasRepository.deleteAll();
    }

    public List<Raqueta> getRaquetas() {
        logger.info("Obteniendo Raquetas");
        return raquetasRepository.findAll();
    }

    public Raqueta createRaqueta(Raqueta raqueta) {
        logger.info("Creando Raqueta");
        return raquetasRepository.save(raqueta);
    }

    public Optional<Raqueta> getRaquetaById(ObjectId id) {
        logger.info("Obteniendo Raqueta con id: " + id);
        return raquetasRepository.findById(id);
    }

    public Raqueta updateRaqueta(Raqueta raqueta) {
        logger.info("Actualizando Raqueta con uuid: " + raqueta.getId());
        return raquetasRepository.save(raqueta);
    }

    public void deleteRaqueta(Raqueta raqueta) {
        logger.info("Eliminando Raqueta con uuid: " + raqueta.getId());
        raquetasRepository.delete(raqueta);
    }
}

File: UD-04/P03-Encordados-MongoDB-JPA-Kotlin/src/main/kotlin/es/joseluisgs/p02encordadosmongodbjpakotlin/controllers/MutuaController.kt
package es.joseluisgs.p02encordadosmongodbjpakotlin.controllers

import es.joseluisgs.p02encordadosmongodbjpakotlin.exceptions.RaquetaException
import es.joseluisgs.p02encordadosmongodbjpakotlin.models.Raqueta
import es.joseluisgs.p02encordadosmongodbjpakotlin.models.Tenista
import es.joseluisgs.p02encordadosmongodbjpakotlin.repositories.RaquetasRepository
import es.joseluisgs.p02encordadosmongodbjpakotlin.repositories.TenistasRepository
import mu.KotlinLogging
import org.bson.types.ObjectId
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Controller

val logger = KotlinLogging.logger {}

// Le inyectamos las dependencias
@Controller
class MutuaController
@Autowired constructor(
    private val raquetasRepository: RaquetasRepository,
    private val tenistasRepository: TenistasRepository
) {

    // TENISTAS

    // Tenistas
    fun tenistasDeleteAll() {
        logger.info("Borrando todos los tenistas")
        tenistasRepository.deleteAll()
    }

    fun getTenistas(): List<Tenista> {
        logger.info("Obteniendo Tenistas")
        return tenistasRepository.findAll()
    }

    fun createTenista(tenista: Tenista): Tenista {
        logger.debug { "Creando tenista $tenista" }
        tenistasRepository.save(tenista)
        return tenista
    }

    fun getTenistaById(id: ObjectId): Tenista? {
        logger.debug { "Obteniendo tenista con uuid $id" }
        return tenistasRepository.findById(id).orElse(null)
        // Mejor lanzar una excepción aqui si es null .orElseThrow { RaquetaException("No existe la raqueta con id $id") }
    }

    fun updateTenista(tenista: Tenista) {
        logger.debug { "Updating tenista $tenista" }
        tenistasRepository.save(tenista)
    }

    fun deleteTenista(tenista: Tenista): Boolean {
        logger.debug { "Borrando tenista $tenista" }
        tenistasRepository.delete(tenista)
        return true
    }

    // RAQUETAS
    fun raquetasDeleteAll() {
        logger.info("Borrando todas los raquetas")
        raquetasRepository.deleteAll()
    }

    fun getRaquetas(): List<Raqueta> {
        logger.info("Obteniendo Raquetas")
        return raquetasRepository.findAll()
    }

    fun createRaqueta(raqueta: Raqueta): Raqueta {
        logger.debug { "Creando raqueta $raqueta" }
        raquetasRepository.save(raqueta)
        return raqueta
    }

    fun getRaquetaById(id: ObjectId): Raqueta? {
        logger.debug { "Obteniendo raqueta con uuid $id" }
        return raquetasRepository.findById(id).orElse(null)
    }

    fun updateRaqueta(raqueta: Raqueta): Raqueta {
        logger.debug { "Actualizando $raqueta" }
        return raquetasRepository.save(raqueta)
    }

    fun deleteRaqueta(raqueta: Raqueta): Boolean {
        logger.debug { "Borrando raqueta $raqueta" }
        raquetasRepository.delete(raqueta)
        return true
    }

    fun findRaquetaByMarca(marca: String): Raqueta {
        logger.debug { "Buscando raqueta por marca $marca" }
        return raquetasRepository.findByMarca(marca)
            ?: throw RaquetaException("No se ha encontrado la raqueta con marca $marca")
    }
}